Add Bellman-Ford algorithm

mstou opened the pull request:
As requested in issue #87, I implemented the bellman-ford algorithm, respecting the coding style of the repository. I included the bellmanFord function in the module exported by lib/alg and added some tests in tests/alg/bellman-ford-tests.js

lutzroeder commented:
@dionyziz can you help reviewing and testing this change?

lutzroeder commented on lib/alg/bellman-ford.js:
```diff
@@ -0,0 +1,74 @@
+
+  var relaxEdge = function(edge) {
+    var edgeWeight = weightFn(edge);
+    if( results[edge.v].distance + edgeWeight < results[edge.w].distance ){
```
`if (results` and `distance) {`

lutzroeder commented on lib/alg/bellman-ford.js:
```diff
@@ -0,0 +1,74 @@
+  var numberOfNodes = nodes.length;
+
+//Relax all edges in |V|-1 iterations
+  for(var i=1; i<numberOfNodes; i++){
```
`for (var i = 1; i < numberOfNodes; i++) {`

mstou commented:
Thanks for the feedback @dionyziz @lutzroeder . Im working on the changes

mstou commented on lib/alg/bellman-ford.js:
```diff
@@ -0,0 +1,74 @@
+  // Get the edges using the edgeFn provided
+  nodes.forEach(function(v){
+    edgeFn(v).forEach(function(e){
+      if(typeof(e)==="object"){
```
Yes, the edge functions should always return an array of edge objects, but inEdges() for example, which is also used on a dijkstra test, returns:
- an empty object if the node has no edges
- an edge object if the node is associated with only one edge 
- an array of edges otherwise

mstou replied:
I stand corrected, edge functions do always return an array of objects. The check here is unnecessary.

mstou commented on lib/alg/bellman-ford.js:
```diff
@@ -0,0 +1,74 @@
+
+  var relaxEdge = function(edge) {
+    var edgeWeight = weightFn(edge);
+    if( results[edge.v].distance + edgeWeight < results[edge.w].distance ){
```
The relaxEdge function is called on the items of edges array, which is created in lines 22-33 where I check If the returned value is an object or an array. In other words, I make sure the function's input is always an edge object. If we change the way we are traversing edges, as mentioned above, this will be corrected correspondingly

About traversing the edges backwards, I was about to open an issue on that because I didnt know if it was preferred or not. So, if we call the edge function on a node V, and in a returned edge object, V is the edge.w then I am using the edge as if it was reversed.

mstou commented on lib/alg/bellman-ford.js:
```diff
@@ -0,0 +1,74 @@
+        });
+      }
+    });
+  });
```
I created the edges array using the edge function ( line 23 ) for code clarity, but I can add it in the main loop as you wish

mstou commented:
@dionyziz @lutzroeder  
* I fixed the code format in all files
* I changed the edge traversal as requested and added the relaxAllEdges() function in bellman-ford.js
* I also corrected the behavior of bellman-ford regarding the reversed edges ( lines 37-40 )
* I merged the common test cases of bellman-ford and dijkstra in  shortest-paths-tests.js

dionyziz commented:
@pkakelas could also help with this review

pkakelas commented:
LGTM, good job :)

mstou commented:
This pull request now fixes #87 .
 Summary:

- Implemented the bellman-ford algorithm
- Created `shortestPaths()` which in Θ(E) checks whether the graph contains a negative weight edge and chooses to use dijkstra or bellman-ford accordingly
- Added the `shortest-paths-test.js` file, which tests the `shortestPaths` function and also exports a test function for shortest paths algorithms that is used in bellman-ford and dijkstra tests to avoid code duplication

assafsun commented:
Hi @mstou 

Thanks for the work!

I'm trying to review the open PR in order to reduce the amount of issues and PRs.
In case you want to continue your work, I will appreciate if you can rebase and create a new iteration. 
I will review it and try to make sure it will be pushed as I think this is a nice algorithm to support.

If not, I will try to continue it.

mstou commented:
Hi @assafsun, I'll be glad to rebase this to master and squash the commits

assafsun commented:
@mstou Thanks for the update, I will follow this PR.

mstou commented:
Hi @assafsun, sorry for the delay, I rebased to master and squashed the fixup commits. Take a look whenever you have time :)

assafsun commented on lib/alg/index.js:
```diff
@@ -8,6 +9,7 @@ module.exports = {
   postorder: require("./postorder"),
   preorder: require("./preorder"),
   prim: require("./prim"),
+  shortestPaths: require("./shortest-paths"),
```
It looks good that you extract it but the documentation will need to mention that this API can run bellman ford or Dijkstra

assafsun commented on test/alg/shortest-paths-tests.js:
```diff
@@ -0,0 +1,106 @@
+
+module.exports = tests;
+
+function tests(algorithm) {
```
Maybe adding a comment that this reflect for bellman ford or Dijkstra?

assafsun commented:
@mstou Thank you for the new iteration. I ran your code and the tests passed.
I left some small comments.

sajaddp commented:
Personally, I really like this pr to get the final approval.
Bellman-Ford algorithm is really practical. Is there an update about this PR or is it better to provide a new code as a PR?
